lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -2237,29 +2240,45 @@ bool TypeChecker::typeCheckPatternBinding(PatternBindingDecl *PBD,
 }
 
 bool TypeChecker::typeCheckForEachBinding(DeclContext *dc, ForEachStmt *stmt) {
+  auto &Context = dc->getASTContext();
+
+  auto failed = [&]() -> bool {
+    // Invalidate the pattern and the var decl.
+    stmt->getPattern()->setType(ErrorType::get(Context));
+    stmt->getPattern()->forEachVariable([&](VarDecl *var) {
+      if (var->hasInterfaceType() && !var->getType()->hasError())
```

Comment:
```suggestion
      if (var->hasInterfaceType() && !var->isInvalid())
```

Plus a comment explaining that we want to try to preserve the types here if they're not actually invalid.

Reply:
Come to think of it, you don't really need the `hasInterfaceType` anymore.

Reply:
`hasInterfaceType()` is still needed because `isInvalid()` calls `getInterfaceType()` which calls `typeCheckASTNodeAtLoc()` for this `ForEach` statement.

As for `getType()->hasError()` -> `isInvalid()`, I just copied this from here: https://github.com/apple/swift/blob/4f106019d616ab220f660bbf0559360e1bc1cbe0/lib/Sema/TypeCheckConstraints.cpp#L2320-L2326
Should I fix that too?

Reply:
Done `getType()->hasError()`→ `isInvalid()` for both instances.